### PR TITLE
Update dataset/dataset field tagging experience

### PR DESCRIPTION
**Pull request by davidsharp7**
### Problem

The Adding and deleting of tags was not providing an intuitive and seemless experience.

### Solution

Based on the feedback i've changed how the datasets details/tags/info components work by introducing a switch which allows field level tags to be exposed.  Refresh only now occurs when the slide toggle is true.  You can add multiple tags in one go now as well.

Screenshot 2024-04-03 at 16 22 46

Screenshot 2024-04-03 at 16 23 03

Screenshot 2024-04-05 at 16 41 18

One-line summary:

Update dataset/dataset field tagging experience.

### Checklist

- [x] You've [signed-off](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#sign-your-work) your work
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [x] You've updated any relevant documentation (_if relevant_)
- [x] You've included a one-line summary of your change for the [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md#unreleased) (_Depending on the change, this may not be necessary_).
- [x] You've versioned your `.sql` database schema migration according to [Flyway's naming convention](https://flywaydb.org/documentation/concepts/migrations#naming) (_if relevant_)
- [x] You've included a [header](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#copyright--license) in any source code files (_if relevant_)

### CODE REVIEW

File: api/src/main/java/marquez/db/DatasetFieldDao.java [hunk 1]
```diff
@@ -129,7 +129,7 @@ void deleteDatasetVersionFieldTag(
 
   @SqlUpdate(
       "INSERT INTO dataset_fields_tag_mapping (dataset_field_uuid, tag_uuid, tagged_at) "
-          + "VALUES (:rowUuid, :tagUuid, :taggedAt)")
```
Comment (phixMe) [inline api/src/main/java/marquez/db/DatasetFieldDao.java:132]: Can you add a Java language test for this?

File: web/src/store/reducers/datasets.ts [hunk 2]
```diff
@@ -32,7 +32,7 @@ export type IDatasetsState = {
   totalCount: number
   init: boolean
   deletedDatasetName: string
-  refreshTags: boolean
+  refreshTags: string
```
Comment (phixMe) [inline web/src/store/reducers/datasets.ts:35]: Why would this be a string and not a boolean?

Comment (davidsharp7) [reply web/src/store/reducers/datasets.ts:35]: So the thinking is twofold - one to bring it into line with

https://github.com/MarquezProject/marquez/blob/78a191b3cb73a50dc7466103e0588e80c3374042/web/src/components/datasets/DatasetDetailPage.tsx#L94

where we are refreshing based on the change in dataset name.

Second reason was for trouble shooting purposes.  Based on the concatanation of namespace/datasetname/field/tag it was just easier to understand/diagnose what was happening.

File: api/src/main/java/marquez/db/DatasetFieldDao.java (hunk 1 above)
Comment (davidsharp7) [reply api/src/main/java/marquez/db/DatasetFieldDao.java:132]: Cool will add an additional insert which should result in no change.

File: web/src/store/reducers/datasets.ts (hunk 2 above)
Comment (davidsharp7) [reply web/src/store/reducers/datasets.ts:35]: Can always change the dependency to 
```
datasets.refreshTags === true
```
and change the reducer to set refreshTags to False when its updating then True on success.

File: api/src/main/java/marquez/db/DatasetFieldDao.java (hunk 1 above)
Comment (phixMe) [reply api/src/main/java/marquez/db/DatasetFieldDao.java:132]: Thanks

File: web/src/store/reducers/datasets.ts (hunk 2 above)
Comment (phixMe) [reply web/src/store/reducers/datasets.ts:35]: Ok, let's just go with what you have here.

File: web/src/components/datasets/DatasetTags.tsx [hunk 3]
```diff
@@ -119,19 +93,37 @@ const DatasetTags: React.FC<IProps> = (props) => {
...
-      ? addDatasetFieldTag(namespace, datasetName, listTag, datasetField)
-      : addDatasetTag(namespace, datasetName, listTag)
+      datasetField
+        ? addDatasetFieldTag(namespace, datasetName, newTag, datasetField)
+        : addDatasetTag(namespace, datasetName, newTag)
+    }
   }
 
   const handleDelete = (deletedTag: string) => {
+    const index = selectedTags.indexOf(deletedTag)
```
Comment (phixMe) [inline web/src/components/datasets/DatasetTags.tsx:119]: We can do this a little more functionally like:
```tsx
const newSelectedTags = selectedTags.filter((tag) => deletedTag !== tag);
setSelectedTags(newSelectedTags);
```

Comment (phixMe) [reply web/src/components/datasets/DatasetTags.tsx:119]: Maybe let's make this change and I'll merge if that sounds good?

Comment (davidsharp7) [reply web/src/components/datasets/DatasetTags.tsx:119]: Surely can.

Comment (davidsharp7) [reply web/src/components/datasets/DatasetTags.tsx:119]: Could I also do this or does it make it less readable?

```
setSelectedTags(selectedTags.filter((tag) => deletedTag !== tag))
```

**Review by wslulciuc — APPROVED** [review]
🤩

File: api/src/test/java/marquez/api/TagResourceIntegrationTest.java [hunk 4]
```diff
@@ -100,4 +100,38 @@ public void testApp_testDatasetTagFieldDelete() {
     // assert that only SENSITIVE remains
     assertThat(taggedDatasetFieldDelete.getFields().get(0).getTags()).containsExactly("SENSITIVE");
   }
+
+  @Test
+  public void testApp_testDatasetTagFieldConflict() {
```
Comment (wslulciuc) [inline api/src/test/java/marquez/api/TagResourceIntegrationTest.java:105]: ❤️

**Pull request merged**